main/datastore_update.es6.js

```diff
@@ -15,77 +14,337 @@ const path = require('path');
+  logger.warn('Added junction DAO indices');
+} else {
+  logger.warn('No indexed DAO found in junction DAO delegate chain');
+}
+
+//
+// Generic algorithm for data import:
+// (1) In parallel:
+// (1a) Sync and unversion Datastore data;
+// (1b) Load/compute new data;
+// (2) Import new data by put()ing what's changed and remove()ing anything in
+//     unversioned Datastore cache that does not also appear in new data.
+//
+
+function doImport(sync, load, daosArray) {
+  return Promise.all([
+    sync().then(function() {
+      return Promise.all(daosArray.map(function(daos) {
+        return unversionData(daos.sync, daos.cache);
+      }));
+    }),
+    load(),
+  ]).then(function() {
+    return Promise.all(daosArray.map(function(daos) {
+      return importData(daos.import, daos.cache, daos.sync);
+    }));
   });
-};
-
-const imported = global.imported = localImporter.import();
-global.releaseCounts =
-    imported.then(() => verify(localCtx.releaseDAO, datastoreCtx.releaseDAO));
-global.apiCounts = imported.then(() => verify(localCtx.webInterfaceDAO,
-                                              datastoreCtx.webInterfaceDAO));
-global.releaseAPICounts =
-    imported.then(() => verify(localCtx.releaseWebInterfaceJunctionDAO,
-                               datastoreCtx.releaseWebInterfaceJunctionDAO));
+}
+
+//
+// Generic unversion + import algorithms.
+//
+
+function unversionData(syncDAO, cacheDAO) {
+  function unversion(versionedDAO, unversionedDAO) {
+    return versionedDAO.select().then(function(sink) {
+      const array = sink.array;
+      logger.log(`Unversioning ${array.length} ${versionedDAO.of.id} objects`);
+      const cls = unversionedDAO.of;
+      let promises = [];
+      for (let i = 0; i < array.length; i++) {
+        promises.push(unversionedDAO.put(cls.create(array[i])));
+      }
+      return Promise.all(promises).then(function() {
+        logger.log(`Unversioned ${array.length} ${versionedDAO.of.id} objects`);
+      });
+    });
+  }
+
+  return unversion(syncDAO, cacheDAO);
+}
+
+function importData(importDAO, cacheDAO, syncDAO) {
+  const E = foam.mlang.ExpressionsSingleton.create();
+  function putData(srcDAO, cmpDAO, dstDAO) {
+    logger.info(`Computing data changes for ${srcDAO.of.id}`);
+    const sink = foam.dao.ArraySink.create();
+    return srcDAO.select(E.DIFF_ONLY(cmpDAO, sink))
+        .then(function() {
+          const array = sink.array;
+          logger.info(`Pushing ${array.length} ${srcDAO.of.id} to Datastore`);
+          let promises = [];
+          for (var i = 0; i < array.length; i++) {
+            promises.push(dstDAO.put(array[i]));
+          }
+          return Promise.all(promises).then(function() {
+            logger.info(`Pushed ${array.length} ${srcDAO.of.id} to Datastore`);
+          });
+        });
+  }
+  function removeData(srcDAO, cmpDAO, dstDAO) {
+    const sink = foam.dao.ArraySink.create();
+    return cmpDAO.select(E.SET_MINUS(srcDAO, sink))
```

Comment:
I'd probably add a check that prevents deletion of the entire database because nothing was loaded, and maybe because everything was marked as not in the loaded data.